### PR TITLE
original banditpam cleanup PR

### PR DESCRIPTION
**Pull request by markymarkdavis**
I moved the code from the individual methods to the KMedoids class, updated the for loops, added documentation, and added batchsize as an optional parameter.

### CODE REVIEW

File: gam/clustering.py [hunk 1]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+    def _finalize(self, count, mean, m2):
+        """Finding variance for each new mean
+
+        Args:
+            count (int): The number of reference points.
+            m2 (int): The updated mean.
+
+        Returns:
+            variance (int): The variance of the medoids
+            TODO: Update this
```
Comment (Lazea) [inline gam/clustering.py:550]: Forgot to update the output documentation here.

File: gam/clustering.py [hunk 2]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+        variance = (m2 / count)
+        sample_variance = m2 / (count - 1)
+        if count < 2:
+            return float("nan")
+        else:
+            return mean, variance, sample_variance
+
+   
+    def _bandit_search_singles(self, X, dist_func, d_nearest, td, tmp_arr, j, i):
+        """Inner loop for pam build and bandit build functions.
```
Comment (Lazea) [inline gam/clustering.py:562]: Inner loop for both `PAM` and `BanditPAM`? The name suggests this is for `Bandit` only.

File: gam/clustering.py [hunk 3]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+
+        Args:
+            X (np.ndarray): The dataset to be clustered.
+            dist_func (callable): The distance function
+            d_nearest (np.ndarray): The distances for all of the centers
```
Comment (Lazea) [inline gam/clustering.py:567]: Missing `td` parameter in the args here.

Comment (Lazea) [reply gam/clustering.py:567]: Looks like `td` is not being used below at all.

File: gam/clustering.py [hunk 4]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+            d_nearest (np.ndarray): The nearest points to the centers.
+            n_used_ref (int): The number of used references
+            mu_x (np.ndarray): The running mean.
+            sigma_x (np.ndarray): The confidence interval.
+            j (int): The solution ids
+            i (int): The index of the center currently trying to be found.
+
+        Returns:
+            mu_x (np.ndarray): The running mean.
+            sigma_x (np.ndarray): The confidence interval.
```
Comment (Lazea) [inline gam/clustering.py:623]: These are out of order, return is `sigma_x, `mu_x`.

File: gam/clustering.py [hunk 5]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+           Recast as a stochastic estimation problem
+           Run time O(nlogn)
+           https://arxiv.org/pdf/2006.06856.pdf
+
+        Args:
+            X (np.ndarray): The dataset.
+            centers (np.ndarray): The center medoids of the different clusters
+            dist_func (callable): The distance function
+            max_iter (int): Max number of times to check for a better medoid.
+            tol (float): [description]
```
Comment (Lazea) [inline gam/clustering.py:803]: Missing description.

File: gam/clustering.py [hunk 6]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+            verbose (bool): Determining whether or not to print out updates
```
Comment (Lazea) [inline gam/clustering.py:804]: I think these two parameters should have default values.

File: gam/clustering.py [hunk 7]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+                list(product(unselected_ids, range(n_clusters))), dtype="int"
+            )
+
+            n_used_ref = 0
+            while (n_used_ref < n_samples) and (swap_pairs.shape[0] > 1):
+                # sample a batch from S_ref (for init, S_ref = X)
+                idx_ref = np.random.choice(unselected_ids, size=self.batchsize, replace=True)
+
+                ci_scale = math.sqrt((2 * math.log(1.0 / delta)) / (n_used_ref + self.batchsize))
+                np.apply_along_axis(lambda a_swap: self._swap_pairs(X, d, a_swap, dist_func, idx_ref, n_used_ref, mu_x, sigma_x, D, E, Tih_min, "h"), 1, swap_pairs)
```
Comment (Lazea) [inline gam/clustering.py:844]: This line is kinda long, could you reformat it?

File: gam/clustering.py [hunk 8]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
     def predict(self, X):
         raise NotImplementedError()
 
+    def _update(self, count, mean, m2, new_values):
+        """ Batch updates mu and sigma for bandit PAM using Welford's algorithm
```
Comment (Lazea) [inline gam/clustering.py:514]: What would the `mu` and `sigma` be in this function?

File: gam/clustering.py [hunk 9]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+        unselected_ids = np.arange(n_samples)
+        unselected_ids = np.delete(unselected_ids, centers[0:i])
+        # solution candidates - S_solution
+        solution_ids = np.copy(unselected_ids)
+        n_used_ref = 0
+        while (n_used_ref < n_samples) and (solution_ids.shape[0] > 1):
+            # sample a batch from S_ref (for init, S_ref = X)
+            idx_ref = np.random.choice(unselected_ids, size=self.batchsize, replace=True)
+            ci_scale = math.sqrt((2 * math.log(1.0 / delta)) / (n_used_ref + self.batchsize))
+            lmbda = np.vectorize(lambda j: self._looping_solution_ids(X, idx_ref, dist_func, d_nearest, n_used_ref, mu_x, sigma_x, j, i), otypes="O")
```
Comment (Lazea) [inline gam/clustering.py:675]: Long line, reformat.

File: gam/clustering.py [hunk 10]
```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
...
+            #
+            # with reduced number of candidates - run PAM swap
+            # TODO - unify full swaps - like was done with search_singles
+            #
+            print(
+                f"Entering swap with {swap_pairs.shape[0]} candidates...pts used = {n_used_ref}"
+            )
+
+            done = True  # let's be optimistic we won't find a swap
+            Tih = np.apply_along_axis(lambda a_swap: self._swap_pairs(X, d, a_swap, dist_func, idx_ref, n_used_ref, mu_x, sigma_x, D, E, Tih_min, "i"), 1, swap_pairs)
```
Comment (Lazea) [inline gam/clustering.py:877]: Long line, reformat.

**Review by Lazea — CHANGES_REQUESTED** [review]
Left some comments, most around documentation. Also I recommend setting up `black` and `flake8` to help reformat some of the code.

**Comment by Lazea** [comment]
Would be good to add the use of `bandit` method to the mushroom example notebook.

**Review by Lazea — APPROVED** [review]
LGTM

**Review by rreilly10 — APPROVED** [review]

**Pull request merged**